packages/libp2p/src/random-walk.ts

```diff
@@ -0,0 +1,130 @@
+import { randomBytes } from '@libp2p/crypto'
+import { TypedEventEmitter, setMaxListeners } from '@libp2p/interface'
+import { anySignal } from 'any-signal'
+import pDefer, { type DeferredPromise } from 'p-defer'
+import { raceSignal } from 'race-signal'
+import type { ComponentLogger, Logger, PeerInfo, PeerRouting, Startable } from '@libp2p/interface'
+import type { RandomWalk as RandomWalkInterface } from '@libp2p/interface-internal'
+
+export interface RandomWalkComponents {
+  peerRouting: PeerRouting
+  logger: ComponentLogger
+}
+
+interface RandomWalkEvents {
+  'walk:peer': CustomEvent<PeerInfo>
+}
+
+export class RandomWalk extends TypedEventEmitter<RandomWalkEvents> implements RandomWalkInterface, Startable {
+  private readonly peerRouting: PeerRouting
+  private readonly log: Logger
+  private walking: boolean
+  private walkers: number
+  private shutdownController: AbortController
+  private walkController?: AbortController
+  private needNext?: DeferredPromise<void>
+
+  constructor (components: RandomWalkComponents) {
+    super()
+
+    this.log = components.logger.forComponent('libp2p:random-walk')
+    this.peerRouting = components.peerRouting
+    this.walkers = 0
+    this.walking = false
+
+    // stops any in-progress walks when the node is shut down
+    this.shutdownController = new AbortController()
+    setMaxListeners(Infinity, this.shutdownController.signal)
+  }
+
+  start (): void {
+    this.shutdownController = new AbortController()
+    setMaxListeners(Infinity, this.shutdownController.signal)
+  }
+
+  stop (): void {
+    this.shutdownController.abort()
+  }
+
+  async * walk (): AsyncGenerator<PeerInfo> {
+    if (!this.walking) {
+      this.startWalk()
+    }
+
+    this.walkers++
+
+    let deferred = pDefer<PeerInfo>()
+    const onPeer = (event: CustomEvent<PeerInfo>): void => {
+      deferred.resolve(event.detail)
+    }
+
+    this.addEventListener('walk:peer', onPeer)
+
+    try {
+      while (true) {
+        this.needNext = pDefer()
+
+        const peerInfo = await deferred.promise
+        deferred = pDefer()
+
+        yield peerInfo
+        this.needNext.resolve()
+      }
```

Comment:
> loop forever (until abort or other thrown I assume)

Not exactly, the loop will exit if the consumer stops iterating over the returned generator.

Given this generator:

```js
async function * generator () {
  let val = 0
  console.info('start loop')

  try {
    while (true) {
      console.info('in loop')
      yield val++
      console.info('after yield')
    }
  } finally {
    console.info('out of loop')
  }
}
```

...if I just consume the first value and then break out of the loop:

```js
for await (const val of generator()) {
  console.info('got value', val)
  break
}
```

The output is:

```
start loop
in loop
got value 0
out of loop
```

You'll notice that `"after yield"` is never logged.

> This seems overly complicated and a lot of indirection to get a consumer aware async generator?

It's a multiple-consumer aware generator.  That is, if there's one consumer, we'll advance the generator at the speed that matches the consumer.  If there are multiple consumers the generator will be advanced at the speed of the fastest consumer.

They may share a value if they pull at the same speed, or one may miss values if it pulls slowly.  This is fine because the peers are randomly chosen so the value stream can be non-deterministic.

> Can we change startWalk to getPeer where we use something like it-take to consume only a single peer from getClosestPeers when asked?

I'd rather not because this will create a new generator and a new random query for every consumer.  This is exactly what I'm trying to avoid here because queries like this are expensive.

Reply:
I've added some more comments, hopefully it's clearer now.